### ISSUE TITLE
Start Before Logon (SBL)?

### ISSUE THREAD

**Issue by kq01526** [OpenVPN/openvpn-gui#77]
Hi,

apparently OpenVPN seems to be able to establish a VPN connection before a user logs on to Windows. Apparently this is being done by having OpenVPN running as a service:

https://openvpn.net/index.php/open-source/documentation/howto.html#startup

Establishing a VPN connection before logging on to Windows is especially useful when needing to log on to a domain joined machine (i.e. having to authenticate on a domain controller).

Is OpenVPN also able to offer a prompt for the VPN credentials before logging on to Windows?

Cisco AnyConnect does have that feature, it's called "Start Before Logon" or in short "SBL", see for example:

https://youtu.be/dkwC5lXu-HQ

Windows itself also does offer a similar feature:

https://blog.lan-tech.ca/2012/04/29/connect-to-windows-vpn-at-logon/
https://blog.lan-tech.ca/2013/03/02/windows-8-connect-to-vpn-before-logon/

So, Windows does seem to have a "native" button for this on the log on screen (which is also being utilized by Cisco AnyConnect).

So, would it also be possible to have OpenVPN prompt for the VPN credentials when clicking on that button, so that a user can enter his VPN credentials (username + password) and connect via VPN before logging on to Windows?

Regards

**Comment by selvanair** [comment]
> So, would it also be possible to have OpenVPN prompt for the VPN credentials when clicking on that button, so that a user can enter his VPN credentials (username + password) and connect via VPN before logging on to Windows?

This requires writing a credential provider that implements the Pre-Logon Access Provider (PLAP) usage scenario. A very basic implementation is not hard, but integrating with service (or with the GUI, or keep it as another way of starting the connection ?), providing options to choose which config to connect etc. etc. takes some work, so may not happen anytime soon. Also such a feature is often not required for domain login: as Windows caches domain login credentials, the user can login and then start the tunnel even if the DC is not reachable before the tunnel is up.

However, this "start before logon" feature may be a needed for setups where the corporate policy does not allow caching of domain login credentials.

**Comment by kq01526** [comment]
> @selvanair wrote:
> 
> Also such a feature is often not required for domain login: as Windows caches domain login credentials, the user can login and then start the tunnel even if the DC is not reachable before the tunnel is up.

That's not quite correct.

For the credentials to be cached, you first need to be able to log on to the domain controller for Windows to be able to cache the credentials.

So you need to log on at least once before the credentials are being cached. In some circumstances this might not be possible without VPN though. So, a Start VPN Before Logon feature would be essential in that case.

Furthermore, the crendential cache storage of Windows is finite.

> @selvanair wrote:
> 
> However, this "start before logon" feature may be needed for setups where the corporate policy does not allow caching of domain login credentials.

Not just that. As Cisco and Microsoft put it:

> http://www.cisco.com/c/en/us/support/docs/security/asa-5500-x-series-next-generation-firewalls/107598-sbl.html
> 
> Start Before Logon lets the administrator control the use of login scripts, password caching, mapping network drives to local drives, and more.

> https://support.microsoft.com/en-us/kb/172931
> 
> When you logon to Windows by using cached logon information, if the domain controller is unavailable to validate your account, you cannot access network resources that require domain validation. However, you can access network resources that do not require domain validation. 

So, it's useful for lots of other things and basically essential for a proper log on.

Furthermore, a Start VPN Before Logon feature would be even more useful now with Windows 10, because Windows 10 even lets you connect to wireless networks before logging on to Windows now.

Regards

**Comment by chipitsine** [comment]
there's similar implementation  https://github.com/Amebis/GEANTLink
we can learn from it

**Comment by selvanair** [comment]
What is blocking this is lack of time not lack of interest.. The part requiring interaction with logonexec is a rather simple COM interface implementation; the rest is starting the VPN, doing user auth etc. which is tedious unless we re-use existing GUI code. I've an implementation that is not yet ready for prying eyes. Need some quality free time to make it decent, sigh...

I think the link you posted does not implement PLAP interface which is required for starting VPN --- as this will execute as a part of logonexec usual auth methods are given a limited time window to complete unless PLAP interface is implemented.

**Comment by sekamin322** [comment]
Does anyone have a definite working solution for starting OpenVPN before login? Switching to AD, our DC is connected to the VPN so if users are offsite and forget their passwords, they will be locked out. Allowing them to connect to the VPN first seems to be the easiest solution.

**Comment by chipitsine** [comment]
what kind of auth do you use ? if login/password, it will not be any usefull.

**Comment by chipitsine** [comment]
I think, computer active directory certificate would be perfect for such kind of setup.

**Comment by rozmansi** [comment]
We use OpenVPN to connect domain computers to organizations _before_ user logs on.

The OpenVPN is started automatically as a service running as the _Local System_ account. We use the NSSM as the service wrapper for this purpose. The service is configured to depend on _Dhcp_ and _tap0901_ services just like the _OpenVPNService_. Probably, the _OpenVPNService_ could do the job too.

For years, we have been using manually managed certificates to authenticate. But asking employees to go get a new certificate to the Windows Server CA _before_ the current one expires, install it to the Local System's Certificate Store, then update the `--cryptoapicert` hash in the _ovpn_ file simply did not scale.

Therefore, we no longer use client certificates but `--verify-client-cert none` + `--auth-user-pass` to authenticate. The server side has a simple script (I can provide samples for Linux and Windows) with a list of approved computer names and a long secret string per each computer. On the client side the `--auth-user-pass` points to an ASCII file with computer name and its secret string. This file has ACL limited access only to _Administrators_ and _Local System_.

Remember, by doing "start-before-logon" you are actually connecting and authenticating _computers_, not _users_.

The end result is, the remote computer is connected automatically soon enough in the Windows boot not only to allow domain logons, but it also processes the Group Policy (when the bandwidth is sufficient). This was the whole point of our setup: the employee's laptops can roam around for years without the need to come back home for service-keeping. Anywhere they go, they are connected from boot.

**Comment by rozmansi** [comment]
A side note: If you use Windows provided VPN to connect before logon, you can use the domain logon - true. However, the Group Policy agent service has tried to contact AD long before that and gave up. Therefore, applying Group Policy settings to the remote computer is not possible.

I am not familiar with Cisco AnyConnect. Can it authenticate and connect soon enough in the boot process so the Group Policy settings can apply too?

**Comment by chipitsine** [comment]
@rozmansi, Active Directory is a combination of DNS, LDAP and kerberos working together.

As long as computer can reach DNS, AD should work. Including domain logon and group policy. I'd like to see debug of group policy service in your case

As for "logon before connect", it is intended to authenticate users, not computers (Your case with computer authentication via service is worth documenting however)

In order to use start before logon, some api should be implemented, api is available starting win8.1 (it was drastically changed in win10)

On win7 Cisco any connect emulates such api by itself

**Comment by rozmansi** [comment]
Here are the exported Group Policy and System event logs:
[SR5.EventLog.boot-on-LAN.zip](https://github.com/OpenVPN/openvpn-gui/files/1436849/SR5.EventLog.boot-on-LAN.zip)
[SR5.EventLog.boot-on-OpenVPN.zip](https://github.com/OpenVPN/openvpn-gui/files/1436850/SR5.EventLog.boot-on-OpenVPN.zip)

Methodology used:
- Clear event logs
- Reboot
- Login
- Export event logs

One run represents laptop connected to the domain LAN. Another run was made using OpenVPN over WiFi with no direct access to domain LAN.

Comparing the event logs you shall notice they are mostly indistinguishable.

**Comment by gordonfrei** [comment]
@rozmansi I would like to try the same with openvpn connecting before logon to get software install from GPO working on remote sites. Can you provide sample script of your setup as mentioned?

**Comment by rozmansi** [comment]
Due to popular demand, I might make a wiki or blog page with step-by-step instructions. Here's something to get you going...

1. Make sure you can connect using an _elevated_ command line:
   ```
   cd "C:\Program Files\OpenVPN\config"
   "C:\Program Files\OpenVPN\bin\openvpn.exe" --config Contoso.ovpn
   ```
   Then disconnect.

2. If you are using `--cryptoapicert`, install the client certificate to the Local Computer's "Personal" store. 

   Optional: I am using `--client-cert-not-required` and username/passwords instead of client certificates, because PKI is a royal pain-in-the-... = you will notice your client certificate expired when you're out on the road.

3. Download and save `nssm.exe` utility. No setup required. For this walkthru, I will assume you copied it to `C:\Program Files\OpenVPN\bin`.

4. Execute:
   ```
   "C:\Program Files\OpenVPN\nssm.exe" install OpenVPN-Contoso "C:\Program Files\OpenVPN\bin\openvpn.exe"
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso AppDirectory "C:\Program Files\OpenVPN\config"
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso AppParameters "--config Contoso.ovpn"
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso AppStdout "C:\Program Files\OpenVPN\log\Contoso.log"
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso AppStderr "C:\Program Files\OpenVPN\log\Contoso.log"
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso AppStdoutCreationDisposition 2
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso AppStderrCreationDisposition 2
   "C:\Program Files\OpenVPN\nssm.exe" set OpenVPN-Contoso DependOnService Dhcp tap0901
   "C:\Program Files\OpenVPN\nssm.exe" start OpenVPN-Contoso
   ```

   This will create an `OpenVPN-Contoso` system service, configure it to run your OpenVPN connection and redirect the output to `C:\Program Files\OpenVPN\log\Contoso.log`. The _...CreationDisposition 2_ tells the `nssm.exe` to make a new log on each service restart. The dependencies will allow correct service startup/shutdown timing. Finally, it will start the service.

You operate VPN connection by starting/stopping/disabling the `OpenVPN-Contoso` service. When set to start _Automatically_, it will connect on boot. The Group Policy Client service does wait for a while on boot to see if any DC becomes available.

Optional: For laptops, I also install and create a Stunnel connection to my VPN server to port 443 to run a secondary OpenVPN TCP connection service thru it (I set its network interface higher metric than its UDP equivalent). This OpenVPN TCP over TLS/443 penetrates the most strict DPI firewalls and brings quite some "surprise" looks when my laptop seamlessly connect to my work in _any_ conference room. :)

**Comment by danieldjewell** [comment]
> A side note: If you use Windows provided VPN to connect before logon, you can use the domain logon - true. However, the Group Policy agent service has tried to contact AD long before that and gave up. Therefore, applying Group Policy settings over VPN is not possible.
> 
> I am not familiar with Cisco AnyConnect. Can it authenticate and connect soon enough in the boot process so the Group Policy settings can apply too?

I can confirm that at least with AnyConnect, connecting the VPN before logon does indeed provide "live" AD logon (i.e. not using cached credentials) and normal group policy processing - as if you were locally connected. If memory serves, there are a few GP settings that can both positively and negatively affect GP processing when using VPN before Logon -- I vaguely remember having to set something like "wait for group policy processing to finish on slow links" or something like that. (It was one of those one-time set it and forget it fixes, so I don't remember the exact steps... if you're experiencing GP processing problems, check those settings ) 

But, in general, without a doubt: AnyConnect definitely has the *capability* of doing full 100% "normal" logins without cached credentials and with full group policy processing (including the tricky ones that will **only run** during a logon.) Likewise, the built-in Windows VPN client also has this ability. 

That said, and without sounding "complain-y", I would stress that the capability of doing this is ***huge*** -- especially for roadwarrior users or users who are 99-100% remote but use domain joined laptops. There are certain policies and settings that *will never* be applied unless they run during an interactive logon (e.g. not using cached credentials). Likewise, as noted elsewhere, the cached credentials have a finite expiration time (as I recall, that's configurable in Group Policy - Security Settings? - somewhere near where one configures password complexity) ... performing an interactive logon refreshes these credentials. 

Unless something has changed with newer releases of Windows 10, logging on to a VPN connection **after** logging into Windows does *not* perform **all** of the same actions as logging in before. Yes, **some** group policies will background refresh and some other things will still work, but if you have a user that is *100%* remote, working from home, doesn't have say a hardware VPN router, eventually, that computer will eventually require an interactive "online" logon. In one of the worst cases, an employee had to ship their laptop across the country overnight so it could be plugged into the network and updated (because no VPN before logon). 

From experience, the AnyConnect client has a setting you enable in the client config to enable the VPN option before logon. Then, in the Windows logon GUI, it would launch the normal client software window "on top of" the logon screen where the user could then interact with it with 100% normal functionality. 

Presumably, since OpenVPN now uses a system service and the client essentially communicates with that service, a similar setup could be used?

**Comment by danieldjewell** [comment]
> You operate VPN connection by starting/stopping/disabling the `OpenVPN-Contoso` service. When set to start _Automatically_, it will connect on boot. The Group Policy Client service does wait for a while on boot to see if any DC becomes available.

But this doesn't provide a mechanism to allow for user authentication or say 2FA? (I mean, doing client certs on smart cards/keys with PINs could even count as 2FA.) If it's running as a service it would seemingly have to use some kind of stored authentication?

What happens if say on boot it doesn't initially connect (say someone forgot to plug in a wired ethernet cable)... is there any way to trigger a reconnect without a reboot?

> Optional: For laptops, I also install and create a Stunnel connection to my VPN server to port 443 to run a secondary OpenVPN TCP connection service thru it (I set its network interface higher metric than its UDP equivalent). This OpenVPN TCP over TLS/443 penetrates the most strict DPI firewalls and brings quite some "surprise" looks when my laptop seamlessly connect to my work in _any_ conference room. :)

Nice! How's the performance? So stunnel is doing a TCP tunnel? And then it's TCP-in-TCP? (More specifically ((TLS-in-TCP)>TLS-in-TCP) for a double tunnel?) I can definitely see the need for using TCP on 443 to fool strict DPI (trying to rack my brain, can't think of any part of HTTP/HTTPS that would use UDP, so yeah). Since you kinda need the outer tunnel to be TCP, couldn't you use UDP inside for better performance? (Trying to think, TCP-in-TCP is never ideal, but UDP-in-TCP might be worse...)

**Comment by rozmansi** [comment]
> But this doesn't provide a mechanism to allow for user authentication or say 2FA? (I mean, doing client certs on smart cards/keys with PINs could even count as 2FA.) If it's running as a service it would seemingly have to use some kind of stored authentication?

True. 2FA may be performed interactively only.

> What happens if say on boot it doesn't initially connect (say someone forgot to plug in a wired ethernet cable)... is there any way to trigger a reconnect without a reboot?

That's quite usual when you walk into a place you have never been before and your laptop doesn't have access to WiFi yet. GP client service will stall the boot for 30 seconds (by default). After that boot will resume offline.

It depends on your .ovpn config. Fortunately, with some `persist...` settings it should keep reconnecting until the network is available.

> Nice! How's the performance? So stunnel is doing a TCP tunnel? And then it's TCP-in-TCP? (More specifically ((TLS-in-TCP)>TLS-in-TCP) for a double tunnel?) I can definitely see the need for using TCP on 443 to fool strict DPI (trying to rack my brain, can't think of any part of HTTP/HTTPS that would use UDP, so yeah). Since you kinda need the outer tunnel to be TCP, couldn't you use UDP inside for better performance? (Trying to think, TCP-in-TCP is never ideal, but UDP-in-TCP might be worse...)

The performance was a bit slower than direct OpenVPN UDP, but still comparable: internet links were the true bottleneck. I made sure the stunnel was up-to-date, using a strong cipher, client-side checking server cert issued by the pinned root CA, and finnaly set OpenVPN to `cipher none`.

However, I never tried to tunnel UDP thru the stunnel.

**Comment by Gauss23** [comment]
Is there any chance of a SBL implementation for OpenVPN? Especially the version, where a box pops up and lets the user enter username and password (+optional 2FA token).
Don´t like the autostart way for various reasons: if the user is in the office, the connection is not needed and for security reasons it´s always good to have multiple separated login factors.

**Comment by cron2** [comment]
Hi,

On Mon, Sep 07, 2020 at 07:48:53AM -0700, Gauss23 wrote:
> Is there any chance of a SBL implementation for OpenVPN? Especially the version, where a box pops up and lets the user enter username and password (+optional 2FA token).

"someone would have to do it"...

I am willing to review and test, but have no idea how such an 
implementation would have to be done.

gert

-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             gert@greenie.muc.de

**Comment by smartens** [comment]
> Due to popular demand, I might make a wiki or blog page with step-by-step instructions. Here's something to get you going...
@rozmansi 
Your setup sounds promising. I'm about to set up a docker server for our startup. Because we will work in a co-working place but still want some central user management we will have a samba domain controller container to which every laptop needs to log in on start. But such a setup with saved credentials is quite a security risk if not properly configured I would be glad if you could provide a step-by-step instruction :).

**Comment by GutsBlack** [comment]
Hello
The Fortinet client (FortiClient 6.x) also allows to launch a VPN before the login.
If the vpn fortigate is connected to the AD you can authenticate the VPN at the same time as the user.
If this is not the case you will have to enter the VPN ID and then the AD user ID

[https://www.deploymentshare.com/using-forticlient-vpn-with-hybrid-aad-logon-after-autopilot-process/](url)

My client abandoned OpenVPN + Pfsense for this reason.

**Comment by mattock** [comment]
I wonder if OpenVPN Connect supports this feature. @lstipakov any idea?

**Comment by GutsBlack** [comment]
OpenVPN Connect and Viscosity do not support this feature either. Viscosity indicated a planning for the feature but it was in 2019...

**Comment by cron2** [comment]
Hi,

On Wed, Jun 30, 2021 at 04:56:46AM -0700, GutsBlack wrote:
> The Fortinet client (FortiClient 6.x) also allows to launch a VPN before the login.

Starting the VPN before login is not convenient but doable, using the
OpenVPN Service (openvpnsrv2).  This only works for VPNs that do not 
require user auth, though.

We have no support for "the VPN starts before login and asks for credentials"
(and I have no idea how to implement that).

But indeed, that would be a truly nice feature.

gert
-- 
"If was one thing all people took for granted, was conviction that if you 
 feed honest figures into a computer, honest figures come out. Never doubted 
 it myself till I met a computer with a sense of humor."
                             Robert A. Heinlein, The Moon is a Harsh Mistress

Gert Doering - Munich, Germany                             ***@***.***

**Comment by GutsBlack** [comment]
> (and I have no idea how to implement that). 

maybe it can help you :
Zyxel use "Gina Mode in windows 7" -> https://docs.microsoft.com/en-us/windows/win32/secauthn/winlogon-and-gina
SystoLOCK use PLAP provider -> https://www.youtube.com/watch?v=XUe1Cf6ruZE

and https://docs.microsoft.com/en-us/windows/win32/secauthn/credential-providers-in-windows

other vpn and start before logon : https://www.shrew.net/static/help-2.2.x/html/Shrew%20Soft%20VPN%20Client%20Administrators%20Guide.html?VPNAccessManager1.html

**Comment by gitnewcomer** [comment]
> Hi,
> On Wed, Jun 30, 2021 at 04:56:46AM -0700, GutsBlack wrote: The Fortinet client (FortiClient 6.x) also allows to launch a VPN before the login.
> Starting the VPN before login is not convenient but doable, using the OpenVPN Service (openvpnsrv2). This only works for VPNs that do not require user auth, though. We have no support for "the VPN starts before login and asks for credentials" (and I have no idea how to implement that). But indeed, that would be a truly nice feature. gert -- "If was one thing all people took for granted, was conviction that if you feed honest figures into a computer, honest figures come out. Never doubted it myself till I met a computer with a sense of humor." Robert A. Heinlein, The Moon is a Harsh Mistress Gert Doering - Munich, Germany ***@***.***

Hi @cron2, this project should help implementing this feature: https://github.com/phaetto/windows-credentials-provider
I think all that is needed is described on Appendix C on this document: https://github.com/phaetto/windows-credentials-provider/blob/master/Credential_Provider_Technical_Reference.pdf
Here we can find more information about the credential provider: https://github.com/DavidWeiss2/windows-Credential-Provider-library
Here another project for older Windows versions (Win7): https://github.com/naveedmurtuza/CredentialProviders
Maybe @lstipakov can help us on that... :)
I would be a great improvement for OpenVPN if this would be supported! Let's make it happen! :)

**Comment by ManZosh** [comment]
In our environment we recently ditched our Cisco ASA which we leveraged the SBL module for VPN connectivity for a Netgate pfSense+ based NGFW appliance. Although this was a real nice upgrade in terms of network performance and licensing overhead, we're now missing out on the awesome functionality and manageability that Cisco's SBL afforded us. We would configure our ASA to our Windows Server's Network Policy Server (NPS) where RADIUS authentication took place, allowing us to configure various network protections and constraints through there. Packaging the SBL module with the end-user's VPN client enabled the Network Sign-In icon at the Windows Logon which gave users an incredibly simple company VPN experience. As Admins we could also push updates and changes to the overall client config and even still the SBL module at connection time would update the client-side config with the new version hosted from the VPN server/ASA.

With SBL, we could always be absolutely sure that all laptops both on-prem and remote were remaining in contact with our Windows DC at logon time, and therefore had various security GPs applied such as refusing PC logon if the domain was unavailable and disabling cached logon credentials. Nowadays, we've had to disable these policies since we haven't been able to get the native Win10 Network Sign-In client to work with the pfSense OpenVPN server... Currently we're evaluating setting up an IPsec-based VPN solution on our firewall as an alternative that seems to probably play nicer with the Win10 native Network Sign-In.

I'd be more than happy to help test out and field feedback to anyone that needs a proper test harness for an OpenVPN version of SBL, as my company sports a full Enterprise network architecture and Windows domain!! Just let me know!! :)

**Comment by gitnewcomer** [comment]
I have made a proof of concept with the CheckPoint Endpoint Security Client so I could test if we could use an OpenVPN connection before Domain Logon...

My steps:

1) Download and install CheckPoint Endpoint Client (I have tested with version E86.00): https://supportcenter.checkpoint.com/supportcenter/portal/user/anon/page/default.psml/media-type/html?action=portlets.DCFileAction&eventSubmit_doGetdcdetails=&fileid=118725
2) Download and install OpenVPN Connect Client: https://openvpn.net/vpn-client/
3) Enable "Secure Domain Logon" on the CheckPoint Endpoint Client options (this will activate the PLAP Provider under HKLM\Software\Microsoft\Windows\CurrentVersion\Authentication\PLAP Providers with GUID {a7fd389f-fac9-4772-b6af-54e09f65a2a3}).
4) Edit the registry key HKLM\SOFTWARE\Wow6432Node\CheckPoint\TRAC\5.0\PRODDIR to point to the OpenVPN Connect installation directory (ex.: C:\Program Files\OpenVPN Connect\).
5) Rename the OpenVPN Connect EXE on the installation directory to the CheckPoint GUI EXE that would be started when we do a Pre-Logon (OpenVPNConnect.exe => TrGUI.EXE).
6) Restart the computer and at Logon, try to click only once on the Network Logon icon: the OpenVPN Connect Client will popup and ask you to import a profile (OpenVPN Client will be run under SYSTEM user, so it will not find any profile that you imported under another user).
7) Connect to the OpenVPN Server: should connect without problems
8) Try to logon on your domain: the CheckPoint PLAP Provider will popup saying it will try to establish a secure connection, you can cancel it and logon will take place.

All this is a proof of concept and not really stable for production! Some problems I found:

1) If you click more than once on the Network Logon icon, you will open every time a new instance of the OpenVPN Client and make it crash.
2) Logon screen may return back to "Hit Ctrl-Alt-Delete to Logon" while still having the OpenVPN Client open.
3) After logon you can't control the VPN connection because the OpenVPN Client is running under the SYSTEM user.
4) You can see that even CheckPoint Client isn't secure enough an let you use another EXE on Network Logon. They should check a hash from the EXE or something before starting it!
5) Maybe more I can't remember right now... :)

So this is something that is doable... but I do not have the knowledge to do it...

Hope someone can make this happend so many admins out there would jump for joy! :P

**Comment by selvanair** [comment]
@gitnewcomer I had worked on this on and off over the years, but I've been holding-off to release it for several reasons including (i) IMO, PLAP UX is not very user friendly (ii) as these dlls are loaded by LogonUI before logon, any bug can badly lockup the machine etc.

I'm also very reluctant to put a major chunk of the GUI-code loaded into a dll running as system -- I know many commercial implementations out there run the whole client GUI process from the logon screen, an approach that is flawed in my view.

By the way, please do not post links to sample implementations unless they are compliant with GPL v2 or later. I'm not  a licensing expert, but I avoid MS samples as some of those are released under MS-PL which is not GPL compatible.

Now, with all those caveats out of the way,  here is an implementation to test out and provide feedback. Implementing this based on MSDN docs is not easy, so use at your own risk.

[PLAP module for OpenVPN](https://drive.google.com/file/d/10eCYdb5qFCE92nQOYyNp7sgh9xkgZPhN/view?usp=sharing)
See the README in the zip (link above) on how to test.

After logging in, if you want the GUI to show the connection, use openvpn-gui.exe from build-artifacts of PR #512. Currently released GUI versions have no ability to connect to prestarted instances.

**WARNING**: A crash in the plap dll will lock you out of the system. You may need to reboot into safe mode and edit the registry to get back in, or have some other backup plan.

**Comment by selvanair** [comment]
fwiw a screenshot of the plap-login screen
![ovpn-plap1](https://user-images.githubusercontent.com/3981391/179145189-8aa48d8f-faa2-4bd1-b15f-648890b2aa25.jpg)

**Comment by gitnewcomer** [comment]
Hi @selvanair,

Thanks for the PLAP module, I will test it and give you a feedback ASAP!
Do you plan to release the source code for the PLAP module so others can improve it / bug fixing it? :)

**Comment by gitnewcomer** [comment]
Hi @selvanair,

I had the opportunity to test your PLAP module and I found following problems:

- if we take to long to use the PLAP module or the connection attempt is taking a little bit longer, we are taken back to the logon screen with the clock (I'm testing on Windows 11, but Windows 10 should be the same). This does not happen with the CheckPoint Endpoint Security PLAP implementation (it stays on the same screen until we provide username/password/connects).
- if we provide a wrong username/password, it will always attempt to connect to the OpenVPN server with the same credentials until I restart the OpenVPN service or reboot the computer. This is critical because we use OTP+password as authentication. If OTP expires, we can't make the PLAP module to ask again the credential (auth-nocache was added to the ovpn; on the PLAP dialog, choosing disconnect does not change anything).
- I would not create one PLAP entry per OVPN, but just have one PLAP entry called "OpenVPN" and when we choose it, ask us to which OPVN we would like to connect. Some OPVN filenames are too big and it isn't so user friendly.

But it is already a good start! Hope we can improve it so OpenVPN could be used more on Windows Domain environments. :)
Thanks for your great work!

**Comment by selvanair** [comment]
@gitnewcomer  Thanks for testing and great feedback.

On Windows 10, I can stay on the PLAP screen for ever... the screen does switch to showing the clock after being idle for 30 seconds or so, but a mouse-click anywhere takes back to the PLAP screen, not the login screen. This is idential to what happens to the native login screen and what the OS wants to do. Probably that's what you are seeing too? The reason why this doesn't happen with Checkpoint is related to the last point below -- not a good thing to emulate, IMO. More on that below.

> if we provide a wrong username/password, it will always attempt to connect to the OpenVPN server with the same credentials until I restart the OpenVPN service or reboot the computer. This is critical because we use OTP+password as authentication. If OTP expires, we can't make the PLAP module to ask again the credentials (auth-nocache was added to the OVPN file. On the PLAP dialog, choosing disconnect does not change the behavior).

This, I do not understand -- the module doesn't change the way openvpn operates. By default (i.e., auth cached, no auth-token et,c), if you give it a wrong username/password that fails to authenticate, openvpn.exe will erase it and prompt again. If the connection succeeds once, it will continue using those credentials without prompting even again. With auth-nocache, it will always prompt for username password whether it was correct or incorrect. 

If you are using OTP, openvpn.exe will always prompt even without auth-nocache, assuming you are doing it the "right" way: i.e., have static challenge in config.  There are many other ways of handling 2-factor auth, but in call cases all that the PLAP cares is whether the daemon prompts for information or not.

In short, whenever the daemon (openvpn.exe) prompts for a password, PLAP module should popup a dialog -- it doesn't cache anything. If that's not happening it's a bug.. I'll do some tests.

> I would not create one PLAP entry per OVPN file, but just have one PLAP entry called "OpenVPN" and when we choose it, ask to which OPVN we would like to connect. Some OPVN filenames are too long and it isn't so user friendly.

I'll give it some thought .... The design of Credential Provider interface allows for multiple credential tiles (connection profiles in our case) with control passed into the module only when the submit button for **a tile** is clicked. This leaves LogonUI in control most of the time and the module takes over only during the `Connect` operation associated with a single connection (tile). I think some implementations violate that; instead, their module behaves as if there is only one connection, chooses it by default  and takes over by moving into `Connect()` from the pov of LogonUI. This may be what Checkpoint Endpoint does, but I'm not keen on copying that kind of high-handed approach. I want to leave LogonUI in control except for short periods of time required to establish a single VPN connection. The difference in behaviour when user sits on the PLAP screen also may be related to this. 

In fact, I'm leaning towards removing even more UI elements from the module to make it leaner and use native PLAP features for collecting user input. The idea is to reduce the code foot-print that executes on the **secure desktop** -- we worked hard to make it possible to run the whole openvpn.exe and the GUI as limited user. A module that runs a major chunk of the GUI code as `SYSTEM` is not something I would want to implement.  That said, let's have more discussion of this.

By the way, long connection names lead to a poor UX in OpenVPN-GUI too. The solution to that would be not to use long connection names, or implement a way to define short names for profiles instead of using their file names.

**Comment by gitnewcomer** [comment]
@selvanair I have to admit that my tests were done in a rush on a remote client. Tomorrow I will be again at the office and will make this tests again and report back...
Thanks again for all your work you've already done! :)

**Comment by selvanair** [comment]
@gitnewcomer In that try to use the latest zip (at the same link) -- I had made some changes which might have fixed some issues that you noticed. Anyway, with auth-nocache I get a prompt every time after a disconnect using the PLAP disconnect button. In fact I''ve removed disconnect/reconnect buttons in the popup as they don't work well in this situation -- has to do with some timing issues with the way openvpn handles signals. Disconnecting using the PLAP disconnect icon is better.

If you copy over the new dll, change the path in registry to the full path to avoid Windows picking up a cached copy. Or reboot.

**Comment by gitnewcomer** [comment]
@selvanair today I had a little more time to test the PLAP module again. I also made a comparative with the CheckPoint Endpoint, so you can see how both behave (see how user friendly the UX is).

1. PLAP screen with both, OpenVPN and CheckPoint Endpoint.
As we can see, the CheckPoint Endpoint only shows one entry for the PLAP module, also wenn it has more than one profile configured. On OpenVPN, if we have more profiles, the screen would be full of icons representing each profile. The CheckPoint Endpoint also starts the connection after choosing the PLAP module, on OpenVPN we have to make an extra click, after choosing the profile, on the connect sign (->) to start the connection. On the CheckPoint Endpoint it already tries to start the connection when choosing the PLAP module. The UX from CheckPoint Endpoint is also cleaner to be used by an end user: on the same window we can choose which profile we want to use and enter credentials/choose certificate (fields adapt while choosing the profile):

![Logon_PLAP](https://user-images.githubusercontent.com/29470811/179975219-dcecd267-436c-4f60-ac7d-c979afa072c1.png)

![Logon_PLAP_CheckPoint](https://user-images.githubusercontent.com/29470811/179975412-83327824-815d-4c59-b449-fd7131f566c0.png)

2. While connecting, the CheckPoint Endpoint UX is also more end user friendly, it only shows a connection progress bar. If we need details (as shown by default on OpenVPN), there is an "Details" button. The details would only confuse end users (to much technical information there).
 
![Logon_PLAP_CheckPoint_Connecting](https://user-images.githubusercontent.com/29470811/179976919-f67b184b-39f4-42bc-ad34-7b7ec93914ab.png)

![Logon_PLAP_CheckPoint_Connecting_Details](https://user-images.githubusercontent.com/29470811/179977405-7e4ec3bd-258e-408d-bfc3-5523f574b19c.png)

3. When we are already connected to a VPN profile and try a new connection, on the CheckPoint Endpoint we got a message telling us that it is already connected and there isn't any possibility to start another VPN connection on top of the present already running. On OpenVPN, if we choose the connected VPN profile, we got a message telling us it is already connected, but, as the OpenVPN PLAP module shows one icon per profile, we can try to connect to a second VPN profile. The second connection doesn't work an only shows a message that it is trying to connect (I think it is trying to connect to the OpenVPN management server). I think one VPN connection on top of another one while login on is not a common scenario, so maybe this is why CheckPoint Endpoint only let us choose the VPN profile one step further: if one is already connected, the window for choosing profile and enter credentials are not shown.

![Logon_PLAP_CheckPoint_Connecting_Connected](https://user-images.githubusercontent.com/29470811/179979136-6aa02f2c-ea26-49a9-b375-2b282a579cee.png)

![Logon_PLAP_OpenVPN_SecondConnection](https://user-images.githubusercontent.com/29470811/179978223-b8864706-528d-4ae9-af3a-fc1d8ff1df9e.png)

5. On OpenVPN PLAP I noticed that on some messages we have to click on two messages: from the PLAP module and Logon UI (here an example when the connection fails).

![Logon_PLAP_OpenVPN_Failed](https://user-images.githubusercontent.com/29470811/179982145-fb3f61d1-9a9b-4fd7-b5b6-63c5779601cd.png)

6. After connecting with an OpenVPN profile an login on, the OpenVPN Gui shows that all connections are connected/trying to connect. Also IP shown is not complete (missing the last octet).

![Logon_PLAP_OpenVPN_Status](https://user-images.githubusercontent.com/29470811/179982889-ece211a1-135f-45a5-85a2-f628dbc158e2.png) ![Logon_PLAP_OpenVPN_Status2](https://user-images.githubusercontent.com/29470811/179982945-2d77f430-563c-49db-a596-e255bc1af165.png)

7. The problems that I reported yesterday are not happening today, so I think it is fixed for now. On the CheckPoint Endpoint it seems that now, after a period of time, we are presented to the logon screen with the clock, it just doesn't do it while trying to connect.

Hope this helps to improve the PLAP module of OpenVPN!

Thanks again for all your hard work making this happen! :)

**Comment by selvanair** [comment]
@gitnewcomer Thanks for all the details and great feedback. I will consider all of these while making revisions.  Very helful indeed. For now some quick remarks:

As I suspected checkpoint is getting into Connect() at first click or PLAP icon which is a design choice that allows them to have a bare-bones PLAP dll which just spawns their client executable on the secure desktop. 

Apart from my dislike for that style, it has issues. If you have multiple PLAP providers and select one that behaves like that, Windows will keep selecting it automatically and not show any of the other tiles until the next logon session.

Some of the other differences you note are because OpenVPN is not Checkpoint :)  We have a diverse user base some technically saavy, some lay users. We support configs in multiple folders with different behaviours -- config-auto, global config_dir and user's config_dir.  Some users have hundreds of config files from "so-called VPN" operators. That said, PLAP would be aimed at dumb corporate users, and most of the time, in a corporate env, there would be just one profile installed and configured by an admin.

For example, the ability to connect multiple connections is a feature of OpenVPN's automatic service and the GUI. Your second connection didn't work probably because it was not designed to work when the first is connected ---  or uses the same management port for both. I generally have three connections active all started from OpenVPN-GUI. Independent of PLAP, all profiles in config_auto folder are auto-started at boot.  A user/admin puts multiple profiles there only if connecting all of them in parallel is okay and desired. In practice, most end user installations will have just one or no profile in that folder and that's all what you will see on the PLAP screen. In general, if you have two profiles, auto-started or not, the GUI also will allow you to start any or all of them -- if any conflict, those configs should have been designed as just one with multiple connection entries in the config file. Or users should know about the conflict. I think this is an administration issue.

I agree, the status/log window is too much info for many users. We could use a progress bar instead -- it will be a fake (or highly non-linear)  progress bar as the time for the initial response from the server is unpredictable (< 1 sec to almost never) and once the handshake starts, connection usually completes in a second or so.

Will fix the annoying popup message on connection errors requiring multiple clicks.
  
The feature in the GUI to show pre-started instances is very new (under review). It was designed for the current automatic service users. As of now it will show config_auto connections as yellow if the connection is waiting for user input or trying to connect to the management interface.  Those connections should be held in `management-hold`  to be not shown as connecting. Again, this is related to the purpose of configs in config_auto folder: supposed to be always connected; anything else is an error/warning.

We may have to tweak the above behaviour if/when PLAP is enabled. Also, PLAP installations should add `management-hold` into the configs in config-auto, while non-PLAP installations need not. Also, check whether your second connection has a management port conflict.

Incomplete IP is an old "bug" -- there is limited room in tray message and it gets truncated especially with long connection names and ipv6 IPs.  The fix would be to remove those details from the tray message, but I was reluctant to do so because it just works for majority of users with one connection.  Yours is a corner case with one `connected` and one in `connecting` state -- we could improve the message in such cases though.  The correct IP is on the status Window.

Thanks again, for suggestions and test report.

**Comment by selvanair** [comment]
oops...

**Comment by gitnewcomer** [comment]
@selvanair 

> If you have multiple PLAP providers and select one that behaves like that, Windows will keep selecting it automatically and not show any of the other tiles until the next logon session.

I have seen this yesterday, didn't know that this was the problem. I thought it could be a conflict between OpenVPN PLAP and CheckPoint PLAP. Now I know why this is happening. :)

> PLAP would be aimed at dumb corporate users, and most of the time, in a corporate env, there would be just one profile installed and configured by an admin.

I agree, PLAP is only useful for domain users, so it would be on most cases only one profile.

> I agree, the status/log window is too much info for many users. We could use a progress bar instead -- it will be a fake (or highly non-linear) progress bar

It doesn't need to be a progress bar, it also could be a spinner. Isn't it possible to first show the credentials window and, after entering the credentials and hitting "Ok", the connection window with a button to show status/log while connecting? It would be less confusing for the end user. If connection fails, it would activate a new button "Retry". Hitting this button would start all over again (ask credentials while hiding the connection status window. This way, if the connection fails, we can hit the details and check the logs). I was thinking something like this:
![image](https://user-images.githubusercontent.com/29470811/180177475-04d6ac03-794d-48cf-876d-b61e7a777ae6.png)

> We may have to tweak the above behaviour if/when PLAP is enabled.

It isn't urgent, the most critical part is already done: the PLAP module! :)

> Incomplete IP is an old "bug"

It's only a cosmetic problem, so it isn't urgent too.

> Will fix the annoying popup message on connection errors requiring multiple clicks.

Let me know when you've done the changes and I will test it!

When are you planning to release the source code for the PLAP module? Maybe others can help you improve it. :)

Thanks again for everything!

**Comment by ManZosh** [comment]
Hello! I've been following this thread as my organization really has a need to have OpenVPN connections initiate at the Windows logon screen, prior to user sign-in.

I went ahead and downloaded @selvanair's [openvpn-plap.zip](https://drive.google.com/drive/folders/1w69yVjfiPVLMJ8V11rfO5GTP1WhX4r_Y?usp=sharing), and followed through the steps to test this out.

**1.** We use a pfSense+ firewall appliance which acts as our OpenVPN server and has a client export & installation utility, so I just exported a Win-Installer bundled with our working .ovpn profile and put it on a spare freshly-installed/updated Win10 laptop. From pfSense:
![openvpn_pfsenseServer](https://user-images.githubusercontent.com/28317348/184911276-ceef2059-9b51-463e-8c74-0af7c9905f5a.png)
What the client's About shows once installed:
![openvpn_about](https://user-images.githubusercontent.com/28317348/184911793-ecfb7249-ffe8-4336-bbcb-9e4cdee620b4.png)
**2.** Once installed, per the README instructions I moved the profile and its dependent crypto files to the config-auto directory:
![openvpn_folder](https://user-images.githubusercontent.com/28317348/184912436-87e7563f-33e2-45a2-a7f5-264ae4b432c7.png)
**3.** I added the management lines to the profile config options. We already require user auth passwords, the logon user accounts are RADIUS'd to our domain controller.
`management 127.0.0.1 1200`
`management-hold`
`management-query-passwords`
`dev tun`
`persist-tun`
`persist-key`
`data-ciphers AES-256-GCM:AES-128-GCM:CHACHA20-POLY1305:AES-256-CBC`
`data-ciphers-fallback AES-256-CBC`
`auth SHA512-256`
`tls-client`
`client`
`resolv-retry infinite`
`remote public.ip.reda.cted 1195 udp4`
`setenv opt block-outside-dns`
`nobind`
`auth-user-pass`
`ca pleiades-UDP4-1195-ca.crt`
`tls-auth pleiades-UDP4-1195-tls.key 1`
`remote-cert-tls server`
`explicit-exit-notify`
**4.** Copied the libopenvpn_plap.dll into C:\Windows\System32
**5.** Then via running regedit.exe as Administrator, I imported the openvpn-plap-install.reg file.
**6.** Restart the OpenVPN Services
![openvpn_sysServices](https://user-images.githubusercontent.com/28317348/184918307-176c4915-6e5b-4362-9ba2-56628f9f2bdf.png)
**7.** Logout of user session, yet at the Windows Logon screen, the **Network Sign-In** button does not show! Also tried rebooting, still no VPN icon. I think that's what you meant by "Network/PLAP" icon in the readme, right?
![Desktop-BHL023-2022-08-16-11-37](https://user-images.githubusercontent.com/28317348/184920983-46dfd7b0-9247-4f96-abba-9d40663c08d6.png)

I also tried going into Windows Network settings and adding a VPN connection to our firewall appliance to force the **Network Sign-In** icon to show at logon. However the only VPN provider available is the **Windows (built-in)** option. I tried to set it up anyway but since the Windows native VPN client doesn't sport any of the config or dependencies of the OpenVPN profile, it can't ever connect to our firewall's OpenVPN server of course.

So this was my attempt at testing in a real-world enterprise capacity what's been discussed here so far. Was I straight up missing some critical part of this?? Or was some portion of my config way off the reservation here?? Do let me know so we can progress further toward a working prototype. My company would also like to $ponsor this feature, if the option can please be made available!

**Comment by selvanair** [comment]
If you used the zip file that was linked to this thread the reg file assumes the dll is in C:\Windows\System32. If you use GHA artifacts for PR #518 you should use the .reg file included in that PR and the dll be copied to C:\Program Files\OpenVPN\bin. That said, your setup looks fine to me, though I recommend using the binary from PR 518 -- several issues were fixed in there.

For testing, there is no need to logout, lock-screen to get the login desktop should be enough. The only requirements for the PLAP icon to show up are (i) have the dll registered and (ii) have at least one ovpn file in config-auto. You can check whether the dll is registered by looking at the registry entries created by the .reg file. And, check your vpn config by connecting to its management interface via telnet.

If you are not very familiar with OpenVPN and Windows system administration, wait for a release where this feature would become available.

**Comment by gitnewcomer** [comment]
Hi @selvanair thanks again that you took the time to add this feature to the OpenVPN.
I've seen that you already submitted the PLAP for revision (PR #518) and saw that there is also the registry file that needs to be imported to the registry to activate the PLAP.
My suggestion, to make more "professional" would be to add an option on the OpenVPN GUI settings to activate the PLAP (if the user does not have registry change rights, just prompt for admin credentials or use one of the OpenVPN services to change the registry, because they normally run under Local System account). By this method, you can add the correct path to the installed PLAP DLL or just remove it when disabled.
Here an example from the CheckPoint Endpoint Security (this always works without admin rights because the CheckPoint Endpoint Security has a service running under Local System account that can handle it):

![image](https://user-images.githubusercontent.com/29470811/185364540-8c509317-7cb6-4191-93df-6c39bae8093b.png)

**Comment by selvanair** [comment]
Installation of plap dll will be handled after merging pr518.. the reg file in there or instructions in the pr  discussion are not necessarily meant for the end user. For now, it's meant  for testing.

I may not add registration option to the GUI as GUI is to be run as limited user. Some products do this using a service ( eg checkpoint that you mention), but not all admins want users to be able to register credential DLLs, so we may not go that route either.

But we'll provide some way for admins to setup plap. It has to be done only once, will not work without profiles in config-auto, so, a one-time admin setup is unavoidable.

**Comment by ManZosh** [comment]
Hey @selvanair! Following the MSVC instructions on the [build.rst](https://github.com/OpenVPN/openvpn-gui/blob/master/BUILD.rst), was able to output a successful build in VS2022, just hit 2 small snags:
- [`CMakePresets.json:8`](https://github.com/OpenVPN/openvpn-gui/commit/09aa7ef73031b0beeb42f59a70440b09f4568f2a#diff-1712b235e02bd4a3ad866fe26447a6dc5c1b449faeea513f449d244d46d03cf5) just had to change this to have "Visual Studio 17 2022" since CMake kept complaining about "_could not find any instance of Visual Studio._"
- Then when running the **x64-release-ossl3** build preset I was getting a "_error MSB8040: Spectre-mitigated libraries are required for this project_" which was easily fixed by just upgrading to MSVC v143 C++ Spectre-mitigated libraries.

Once those were taken care of I then had a binary built from PR518! Lil screenshot to confirm it's from your PR with all the PLAP goods:
![image](https://user-images.githubusercontent.com/28317348/187294587-fbbcd268-f07f-4488-b449-71c7d2ecf6e4.png)

I basically followed the rest of the above steps as last time. I did notice the `openvpn-gui\plap\openvpn-plap-install.reg` had the DLL location set to OpenVPN's source bin instead of Sys32 now. So I moved the **new binary** along along with the **libcrypto-3-x64.dll** (_presumed this, otherwise an install error yields if not found_) & **libopenvpn_plap.dll** into my install directory at `..\Program Files\OpenVPN\bin`. Added the reg file to registry, and confirmed my firewall's profile (_**with** the management lines added per your readme_) was present in the config-auto folder. I noticed OpenVPN seems to do some kind of caching to %userprofile%\OpenVPN\config, maybe this is due to the config-auto folder? Regardless I removed the duplicate profile items out of there.

After a recycle of the OpenVPN services, I was able to connect to my firewall over a hotspot without any issue! Which is a great confirmation that all of the above seems kosher. The only big problem negating all this is that unfortunately the Network Sign-In icon **does NOT** at any point show on the Windows Logon screen! The laptop I've been using to test is a fresh install of Win10Pro. Even after a reboot, no dice. On logged-in account lock, still no icon, whether the VPN is on and connected or not.

---
Am I doing something wrong here? Is adding the registration for the plap.dll supposed to be all it takes to enable that VPN button on the Login screen? I feel like I'm on the cusp of being the first enterprise user to be on the cutting-edge of OpenVPN with SBL! Let me know what you think!

---
My OpenVPN log after all the above changes and a successful connection from the desktop system tray icon:
`2022-08-29 16:08:25 OpenVPN 2.5.2 x86_64-w64-mingw32 [SSL (OpenSSL)] [LZO] [LZ4] [PKCS11] [AEAD] built on Apr 21 2021`
`2022-08-29 16:08:25 Windows version 10.0 (Windows 10 or greater) 64bit`
`2022-08-29 16:08:25 library versions: OpenSSL 1.1.1k  25 Mar 2021, LZO 2.10`
`Enter Management Password:`
`2022-08-29 16:08:32 TCP/UDP: Preserving recently used remote address: [AF_INET]re.da.ct.edd:port`
`2022-08-29 16:08:32 UDPv4 link local: (not bound)`
`2022-08-29 16:08:32 UDPv4 link remote: [AF_INET]re.da.ct.edd:port`
`2022-08-29 16:08:32 [vpn.realcompany.com] Peer Connection Initiated with [AF_INET]re.da.ct.edd:port`
`2022-08-29 16:08:33 open_tun`
`2022-08-29 16:08:33 tap-windows6 device [OpenVPN TAP-Windows6] opened`
`2022-08-29 16:08:33 Set TAP-Windows TUN subnet mode network/local/netmask = red.act.ed.d/red.act.ed.ddd/255.255.255.0 [SUCCEEDED]`
`2022-08-29 16:08:33 Notified TAP-Windows driver to set a DHCP IP/netmask of red.act.ed.ddd/255.255.255.0 on interface {E887B094-5E15-4C53-8D65-4610F9120982} [DHCP-serv: red.act.ed.ddd, lease-time: 31536000]`
`2022-08-29 16:08:33 Successful ARP Flush on interface [19] {E887B094-5E15-4C53-8D65-4610F9120982}`
`2022-08-29 16:08:33 IPv4 MTU set to 1500 on interface 19 using service`
`2022-08-29 16:08:33 Blocking outside dns using service succeeded.`
`2022-08-29 16:08:38 WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this`
`2022-08-29 16:08:38 Initialization Sequence Completed`
`2022-08-29 16:08:38 Register_dns request sent to the service`

Lastly the locked desktop showing the Network Sign-In button is missing:
![image](https://user-images.githubusercontent.com/28317348/187301561-483a3db6-69ad-4782-aa32-16c70105d959.png)

**Comment by selvanair** [comment]
The GUI does not "cache" any configs in %userprofile%  --- if you have one there you must have put it there (or imported using the GUI menu) at some point. Anyway, such duplicates will only affect the GUI operation, not PLAP. 

The posted log shows a connection started from scratch by the GUI through the interactive-service. So something is not right. The connection should be controlled by OpenVPNService and running independent of the GUI.

Stop the GUI, remove all ovpn files in %userprofile%/OpenVPN/config and below it as well as in C:\Program Files\OpenVPN\config, have at least one profile in the config-auto folder, check whether openvpnservice is running, check task manager to see whether openvpn daemon started by that service is running, use telnet to connect to its management interface, check the log file of the connection started by the service (not from GUI) --- these are some of the things you could do to ensure the profile in config-auto is ready for control by the PLAP dll.

**Comment by selvanair** [comment]
@ManZosh By the way, I have noticed a bug in the dll that could cause it to crash the login screen if config-auto folder is missing. So be careful while testing and know how to get back in via safeboot if you get locked out. I'll fix this, but its unrelated to the trouble you are having.

**Comment by selvanair** [comment]
@ManZosh  Check whether the default value of the registry key HKEY_CLASSES_ROOT\CLSID{4fbb8b67-cf02-4982-a7a8-3dd06a2c2ebd}\inprocserver32` reads `C:\Program Files\OpenVPN\bin\libopenvpn_plap.dll` If  not fix it. I found a bug in the reg file that could have caused this not to be set.

**Comment by bjoernv** [comment]
I am working on a "Start Before Logon" setup where an OpenVPN client should be started before Windows login, but only if a Smart card is inserted.

Currently the problem is, that the Smart card needs a PIN, which can't be entered before Windows login.

I wrote a feature request for an option to save the PIN. See https://community.openvpn.net/openvpn/ticket/1485

An alternative solution would be to extend the Windows service for OpenVPN so that the service uses the OpenVPN management interface to enter the PIN from a configuration file. The current version of the OpenVPN Windows service probably can not do this.

**Comment by selvanair** [comment]
> Currently the problem is, that the Smart card needs a PIN, which can't be entered before Windows login.

Starting verion 11.30 PLAP module is available and interatcive connection of OpenVPN tunnel with password/PIN/challenge response input from login screen is now possible. See #518

**Comment by bjoernv** [comment]
Could you please give a HOWTO for testing? I have OpenVPN 2.5.8 community edition installed.

And I have the compiled the Git master version of OpenVPN-GUI. The openvpn/out/build/x64-release-ossl3/ directory now contains these *.exe and *.dll files:
```
libcrypto-3-x64.dll
libopenvpn_plap.dll
openvpn-gui.exe
test_plap.exe
```
but no openvpnserv2.exe and no installer.

How I test the OpenVPN-GUI PLAP functions?

**Comment by selvanair** [comment]
It may be easier to test this using a snapshot installer for 2.6 -- see [here](https://build.openvpn.net/downloads/snapshots/github-actions/openvpn2/)

Otherwise, re-install 2.5.8 by selecting to install automatic service. It's not selected by default. Set to start the service automaticlaly and then start it. Copy `libopenvpn_plap.dll` that you built to `C:\Program Files\OpenVPN\bin`. Copy `plap/openvpn-plap-install.reg` from the GUI sources to the PC and run it as admin to register the dll. These steps are handled by the installer in snapshot builds (eventually in 2.6 release).

Then follow instructions in README -- first get [persistent connections](https://github.com/OpenVPN/openvpn-gui#persistent-or-pre-started-connections) working from the GUI and then [from PLAP](https://github.com/OpenVPN/openvpn-gui#interactively-starting-connections-before-logon).
Note that you will have to restart `OpenVPNService` if you add or modify config files in config-auto folder.

**Comment by ManZosh** [comment]
@selvanair Hi Selva! I just had a brief question for you, picking back up on all this... I installed on a new Win10Pro laptop a fresh OpenVPN client using the [openvpn-master-20221216T0629-97929d16-amd64.msi](https://build.openvpn.net/downloads/snapshots/github-actions/openvpn2/openvpn-master-20221216T0629-97929d16-amd64.msi) you linked. I confirmed the OpenVPN service is running, alongside the interactive service (_if that's even necessary for this_). My profile is the same as my [above comment ](https://github.com/OpenVPN/openvpn-gui/issues/77#issuecomment-1216825708) as well, with the management lines added. The PLAP VPN button appears on the Windows Logon screen! But for some reason the client gets hung up on waiting for the management network to come up, shown below:
![image](https://user-images.githubusercontent.com/28317348/208537146-be44fa86-dc6b-4706-940d-bc9328244019.png)

I don't recall having this hung management interface previously when I compiled a [client from source](https://github.com/OpenVPN/openvpn-gui/issues/77#issuecomment-1230889186).

Is the management interface literally the TAP-Windows Adapter V9? Trying a variety of trial & error troubleshooting steps, but if the management lines in the config file change at all, the PLAP icon is no longer visible on the Logon screen.

**Comment by selvanair** [comment]
You can test connection to management i/f by logging in and running OpenVPN-GUI. It should show profiles in config-auto grouped under Persistent Connections and will allow you to connect it. If the GUI also fails to connect to the management interface, either the `management` option in the config is wrong or some other program is holding a connection to that port and thus blocking it. It can also be tested by `telnet  127.0.0.1 port` where port is the one specified in the `management` option in the config file.

BTW, now that 2.6_beta is out testing using the latest beta release (26_beta2) may be better.

**Comment by gitnewcomer** [comment]
@selvanair I've been trying the PLAP without management interface password and all worked fine.
Now that I would like to test it on our corporate environment, I have added a password file to the management interface like this:

`management 127.0.0.1 8888 passwd.txt`

The passwd.txt file has one line with the management interface password. I've copied the file to the same path as the ovpn (config-auto), but after this, I can see the PLAP on Windows Logon screen and when I choose the ovpn profile it doesn't ask any username/password and shows that the connection failed.
If I try to start the connection with the OpenVPN GUI then I see a quick window opening and closing again. Log file does not log anything.
Am I missing something or doing something wrong?
Thanks again for all you work for making this PLAP a reality! :)

**Comment by ManZosh** [comment]
@selvanair 
> BTW, now that 2.6_beta is out testing using the latest beta release (26_beta2) may be better.

Alrighty, please ignore my previous comment! Something was funky with that particular laptop's network stack that was preventing the management interface from being built. I tested this on a new laptop and everything works beautifully!! Here are the steps I followed...

1. Cleanly fresh install [OpenVPN GUI v2.7git](https://build.openvpn.net/downloads/snapshots/github-actions/openvpn2/openvpn-master-20221220T0414-e567f342-amd64.msi) (_beta?_), ensure the **Pre-Logon Access Provider** (PLAP) is enabled for install.
2. Once installed, open services.msc and change the **OpenVPNService**'s `Startup Type` to _Automatic_.
    - (_is this an available directive in the .msi installer?_)
3. Ensure the following new configuration lines are appended to our OVPN profile:
    - `management 127.0.0.1 1200`
    - `management-hold`
    - `management-query-passwords`
4. Insert the modified profile and its dependencies into `..\OpenVPN\config-auto`, there should not be a profile in the regular `config` folder unless you have an otherwise need for it.
5. If you sign-out of the user session or reboot, via the newly implemented PLAP there's now a 4th icon available on the bottom-right of the Windows Logon screen which reads **Network sign-in** on mouse hover. Clicking this shows the connection the OpenVPN service built where we can simply click the connect arrow and the user can then sign-in using (_atleast in our case_) their LDAPS credentials.
6. Start the OpenVPNService, on startup this service checks the `config-auto` folder for any connections it should automatically build.

The connection is immediately visible from the OpenVPN server. With this, we can now achieve a Windows SysAdmin's dream of having a laptop asset residing on a foreign untrusted network be available to our Domain Controller and Remote Assistance software agent before the user is even signed into their Windows user profile!

Only 3 small bugs I've noticed through successively testing this...
- At the Windows Logon screen, if you proceed through to Sign-In (_or Ctrl+Alt+Del_) and wait for the default 2-minute timeout where it throws you back to the Lock screen wallpaper, if you proceed through to Sign-In a second or successive time the PLAP icon is no longer available and (_atleast as an end-user_) you can't bring it back unless you reboot.

- If you connect to the VPN via PLAP, the PLAP icon then shows an 'X' on it providing you the obvious ability to now Disconnect your VPN session. However, if you sign-in and then sign-out of the Windows user profile, end up back at the Lockscreen, and proceed through to sign-in again, the PLAP icon no longer has the 'X' anymore and appears as if you can initiate a _new_ VPN connection. But as evident from the OpenVPN Server's client statuses, the VPN connection is still alive and active! If you attempt to click the PLAP again, and hit the connect arrow on your auto-config'd profile, it immediately shows you a Connected message again. This may have potential to confuse end-users.

- At anytime, if the user does decide to use the PLAP icon with an 'X' on it to kill the VPN connection, the icon just then disappears on-click and does **not** give you ability to connect to the VPN again unless you reboot the system.

So this is just incredible and I'm stupidly excited to soon use this in a production capacity in our environment. Please let me know if I can be of any assistance!! Can take GIFs of any of the bugs I mentioned as well if necessary.

**Comment by selvanair** [comment]
@ManZosh 

Thanks for the feedback

> Only 3 small bugs I've noticed through successively testing this...
> 
> * At the Windows Logon screen, if you proceed through to Sign-In (_or Ctrl+Alt+Del_) and wait for the default 2-minute timeout where it throws you back to the Lock screen wallpaper, if you proceed through to Sign-In a second or successive time the PLAP icon is no longer available and (_atleast as an end-user_) you can't bring it back unless you reboot.

This I cannot reproduce and should not happen. Unless the dll crashes which would be a bug. But I have not seen it.

> * If you connect to the VPN via PLAP, the PLAP icon then shows an 'X' on it providing you the obvious ability to now Disconnect your VPN session. However, if you sign-in and then sign-out of the Windows user profile, end up back at the Lockscreen, and proceed through to sign-in again, the PLAP icon no longer has the 'X' anymore and appears as if you can initiate a _new_ VPN connection. But as evident from the OpenVPN Server's client statuses, the VPN connection is still alive and active! If you attempt to click the PLAP again, and hit the connect arrow on your auto-config'd profile, it immediately shows you a Connected message again. This may have potential to confuse end-users.

I think this is a limitation of PLAP. It does not show the connected icon without going through a connect cycle and does not remember that state from session to session. Within a session (even after the 2 min timeout) it should remember it, I believe. I do not think we can do anything about it. If you have seen any other PLAP offerings (such as anyconnect) behave differently, I can take a look again. One thing we could possibly do is to show a connected checkmark on the tile but its not easy to implement, so unlikely to happen any time soon.

> * At anytime, if the user does decide to use the PLAP icon with an 'X' on it to kill the VPN connection, the icon just then disappears on-click and does **not** give you ability to connect to the VPN again unless you reboot the system.

Looks similar/related to your first point. Again, this should not happen and I cannot reproduce. May be you can open a separate issue with details of OS and OpenVPN version etc.

**Comment by selvanair** [comment]
@gitnewcomer Have you seen this mis-behaviour on login-screen time-out or after clicking PLAP disconnect button reported by @ManZosh above? Sounds like a bug in the dll, but I cannot reproduce it.

**Comment by ManZosh** [comment]
> @gitnewcomer Have you seen this mis-behaviour on login-screen time-out or after clicking PLAP disconnect button reported by @ManZosh above? Sounds like a bug in the dll, but I cannot reproduce it.

Hey @selvanair, it's worth mentioning the only difference there could be between your test system and mine is that the laptop I've been using is indeed joined to our production Windows Domain and therefore has many policies applied. Now I'd never think any of our policies could be affecting this, but if in your experience the PLAP icon indeed persists through Logon timeouts, then it could just be something deep in the bowels of our particular Group Policy that might be affecting these PLAP phenomena I'm seeing.

I have other laptops in inventory I can put a totally fresh install of Win10Pro onto, this time I won't join it to my domain at all. I'll just run through the brief setup steps and try again and see if those bugs persist. If not, then it must be something we're flicking on or off in GP on domain-joined PCs.

Lemme get to that after the holiday and I'll be sure to report back here!

**Comment by gitnewcomer** [comment]
> @gitnewcomer Have you seen this mis-behaviour on login-screen time-out or after clicking PLAP disconnect button reported by @ManZosh above? Sounds like a bug in the dll, but I cannot reproduce it.

Hi @selvanair, first of all, I wish you and all here a Happy New Year! :)
Sorry for my late reply, but I was on holiday between Christmas and New Year without access to my git account.
To the problems reported by @ManZosh, I didn't remember having them, but I could test it on my test lab (computer joined to a domain) and report back as soon as I can...

**Comment by MP230** [comment]
Hello Git-OpenVPN-Developer,
I want to try the new "Pre-Logon Access Provider" Feature with the 2.6 RC "OpenVPN-2.6_rc1-I001-amd64.msi".
With a GUI customize installation it works fine when I manual checked "Pre-Logon Access Provider".
But I doesn't find any installation Parameter for an silent MSI istallation.
I tryed it with "select_service=1" without sccess.

Can anyone help me?

**Comment by lstipakov** [comment]
> Can anyone help me?

This should work:

`c:\Users\lev\Downloads>msiexec /i "OpenVPN-2.6_rc1-I001-arm64 (1).msi" /quiet ADDLOCAL="OpenVPN.GUI,OpenVPN.GUI.OnLogon,OpenVPN.Service,OpenVPN.PLAP.Register,OpenVPN.Documentation,OpenVPN.SampleCfg,Drivers,Drivers.OvpnDco,Drivers.TAPWindows6,OpenSSL,EasyRSA" /L*V install.log`

You might want to remove `Drivers.OvpnDco` if you install on pre-Windows 10 system.

**Comment by gitnewcomer** [comment]
> @gitnewcomer Have you seen this mis-behaviour on login-screen time-out or after clicking PLAP disconnect button reported by @ManZosh above? Sounds like a bug in the dll, but I cannot reproduce it.

Hi @selvanair and @ManZosh,

I had some time to test the PLAP again on my test lab (computer is running Windows 11 Pro with domain joined, so Windows 10 Pro could behave differently).
I have used the last OpenVPN beta version from 17/01/2023 found here [OpenVPN Development Snapshots](https://build.openvpn.net/downloads/snapshots/github-actions/openvpn2).

> At the Windows Logon screen, if you proceed through to Sign-In (or Ctrl+Alt+Del) and wait for the default 2-minute timeout where it throws you back to the Lock screen wallpaper, if you proceed through to Sign-In a second or successive time the PLAP icon is no longer available and (atleast as an end-user) you can't bring it back unless you reboot.

I can not reproduce this. If I start a OpenVPN sign-in process (where it asks me username and password from OpenVPN) and wait some time, I will get to the Windows lock screen. If I try to sign-in again to OpenVPN, the PLAP shows again like expected.

> If you connect to the VPN via PLAP, the PLAP icon then shows an 'X' on it providing you the obvious ability to now Disconnect your VPN session. However, if you sign-in and then sign-out of the Windows user profile, end up back at the Lockscreen, and proceed through to sign-in again, the PLAP icon no longer has the 'X' anymore and appears as if you can initiate a new VPN connection. But as evident from the OpenVPN Server's client statuses, the VPN connection is still alive and active! If you attempt to click the PLAP again, and hit the connect arrow on your auto-config'd profile, it immediately shows you a Connected message again. This may have potential to confuse end-users.

On Windows 11 I don't see any "X" on the PLAP icon after successfully connect to OpenVPN. If I click on the PLAP icon again, I only can try to sign-in again an am presented with a message that I'm already connected. I do have a new icon near Network Sign-In that allows me to disconnect the active OpenVPN connection if still connected. Maybe this was improved on Windows 11.

> At anytime, if the user does decide to use the PLAP icon with an 'X' on it to kill the VPN connection, the icon just then disappears on-click and does not give you ability to connect to the VPN again unless you reboot the system.

As I do not see the "X" on PLAP icon on Windows 11, I can't reproduce this behavior. But if I use the new icon to disconnect the active OpenVPN connection I am able to connect again. As said above, this can be an improvement of Windows 11.

If I have some time (having some project running right know at work), I will setup a new computer with Windows 10 Pro on my test lab to test this steps again...

**Comment by MaartenJB** [comment]
Ok nice, Thanks :-)

Ah yes, it's before user logon, Thanks for the detailed answer!

**Comment by virtualizer117** [comment]
This page is freakin' awesome, so thank you (@selvanair) and everyone else for this incredible timeline of progress. If I found this page looking for a solution to accessing OpenVPN prior to logon, I'm sure others will, too, so I thought I'd add my two cents.

I'm using the OpenVPN-2.6.1-I001-amd64.msi release ([found here](https://openvpn.net/community-downloads/)) on a Windows 10 Pro machine (not domain-joined YET -> domain logon without cached credentials is the end goal). The install itself is flawless, and I went and followed the steps outlined by @ManZosh:
> 1. Cleanly fresh install [OpenVPN GUI v2.7git](https://build.openvpn.net/downloads/snapshots/github-actions/openvpn2/openvpn-master-20221220T0414-e567f342-amd64.msi) (_beta?_), ensure the **Pre-Logon Access Provider** (PLAP) is enabled for install.
_=> (I used the 2.6.1 release mentioned above)_
> 2. Once installed, open services.msc and change the **OpenVPNService**'s `Startup Type` to _Automatic_.
_=> (actually defaulted to "Automatic" for me)_
> 3. Ensure the following new configuration lines are appended to our OVPN profile:
>        * `management 127.0.0.1 1200`
>        * `management-hold`
>        * `management-query-passwords`
_=> (@selvanair also suggests adding * `auth-retry interact`, though I think it was mentioned that this is not 100% necessary)._
> 4. Insert the modified profile and its dependencies into `..\OpenVPN\config-auto`, there should not be a profile in the regular `config` folder unless you have an otherwise need for it.
> 5. If you sign-out of the user session or reboot, via the newly implemented PLAP there's now a 4th icon available on the bottom-right of the Windows Logon screen which reads **Network sign-in** on mouse hover. Clicking this shows the connection the OpenVPN service built where we can simply click the connect arrow and the user can then sign-in using (_at least in our case_) their LDAPS credentials.
> 6. Start the OpenVPNService, on startup this service checks the `config-auto` folder for any connections it should automatically build.
_=> (this is the part that I got hung up on--see below)_

===================
@Marvo2011 (different thread) and @ManZosh (this thread) both mentioned/provided screenshots of issues where the OpenVPN PLAP window console window hangs with the message _**"Waiting for the management interface to come up"**_. @ManZosh addresses this issue and actually provides the suggested fix (also referenced by @selvanair [here](https://github.com/OpenVPN/openvpn-gui/pull/518) when he references starting the OpenVPNService.

In my case, the OpenVPNService was already running by default, so I had to restart it in order for OpenVPN to recognize the config file/options properly (note: @selvanair mentioned this was necessary if you copy-pasted the config file into the `config-auto` directory).

That being said, @ManZosh, I know I'm super late to the party, but you said there was an issue with the way the network stack was built on the computer that was preventing the management interface from being built--I'm guessing you would have already restarted the OpenVPNService, so that was actually a different problem, I guess. Did you ever go back to that particular computer, do any sort of wipe, and try it again? Just curious in case I ever come across a similar situation in the future.

I'm running some more tests now; I'll provide any updates if they end up being worthwhile. I'm seriously impressed by how far this has come, so thanks everyone for the amazing effort.

(BTW this is only my third post on GitHub--recent join--so thanks everyone for putting up with my terrible formatting; I will improve :D).

**Comment by selvanair** [comment]
Be very careful if you do any of these things: configs in config-auto or linked to there are started as SYSTEM. If there are scripts, those will also run as SYSTEM allowing users to essentially run anything as SYSTEM.

**Comment by Luke-Williams9** [comment]
Can someone explain quickly how I can test PLAP? I do not see any icon for it on the logon screen. I'm using OpenVPN client 2.6.6, and confirmed that PLAP/gui.onlogon is installed (all components are installed)
I've put a test configuration in the config-auto folder, restarted the services, but there is still no icon. What am I missing?

**Comment by selvanair** [comment]
> I've put a test configuration in the config-auto folder, restarted the services, but there is still no icon. What am I missing?

See this section and the one following it in [README](https://github.com/OpenVPN/openvpn-gui#persistent-or-pre-started-connections )

**Comment by cyacelga** [comment]
Hello, could someone please help me so I can launch the automatic OpenVPN connection from the hidden icon, I would appreciate it.
I have passed the username and password through a txt file, now I require the connection to automatically start when logging into Windows 11.
thank you
start .\openvpn-gui.exe --silent_connection 1 --connect archivo_config.ovpn
don´t found this line, please help me.

**Comment by selvanair** [comment]
> start .\openvpn-gui.exe --silent_connection 1 --connect archivo_config.ovpn
don´t found this line, please help me.

Instead asking an unrelated question in a closed issue, open a new issue explaining what you are trying to achieve. I do not see the need for what you are doing as the GUI can be start automatically on login and connections will autostart if you do not actively disconnect on logout.

**Comment by gitnewcomer** [comment]
Hi @selvanair, time has passed and your PLAP plugin is still working fine until today! 😃 

Because of change on our security plans, we are now switching to OpenVPN with OTP.
On new OpenVPN GUI there is an option to show a new field for the OTP when logging in.

![Image](https://github.com/user-attachments/assets/cb85739f-09d7-44e3-bbd7-8293e31a6369)

Could you replicate that on the PLAP plugin? I know the user can add OTP at the end of the password field, but we need to make it a little bit more user friendly... 😃
There is also a possibility on OpenVPN GUI to save username and password: could you implement this on PLAP plugin **ONLY** when the OTP option is active?

![Image](https://github.com/user-attachments/assets/a7d312ef-6022-45d6-843c-be4cc65705a0)

With this change, the user would only need to enter username and password once and in the future only the OTP code (this OTP code changes every 30 seconds, so saving username and password would be, on our perspective, not a security flaw, and for those that don't want to use it, can toggle the settings, like on OpenVPN GUI).

Thanks again for you patience and time for making this plugin a reality! 😃

**Comment by selvanair** [comment]
@gitnewcomer  Thanks for the feedback.

In OpenVPN 2.7 you can specify `static-challenge  "Input OTP" concat` [1] and continue using the current setup. As 2.7 release is around the corner (hopefully), can you wait for that?
 
As for saving password in PLAP mode, exposing a password on the login screen looks just wrong.  We can implement saving the previously used username.

If password security is not important it could be hardcoded in a file and passed to `auth-user-pass` directive in the config file. Or, even inlined username/password is now supported:
<auth-user-pass>
username
password
<auth-user-pass>

---
[1] Replace "Input OTP" by a prompt that should be shown to the user.

**Comment by gitnewcomer** [comment]
> In OpenVPN 2.7 you can specify `static-challenge "Input OTP" concat`

@selvanair do you mean adding this to the OVPN config file and it will show the OTP field on PLAP?

> As for saving password in PLAP mode, exposing a password on the login screen looks just wrong.

On login screen there isn't any clipboard functionality (from what I know, all remote support tools send keystrokes on logon screen), how would it be possible to read the saved password? And if there is a chance to read it, it won't give you access to the VPN without the OTP, this is why I mentioned it should only be possible when the option for showing the OTP field is active. Or where do you think it could be a security problem?

> We can implement saving the previously used username.

This would be saved automaticaly if the "save password" would be available, right?

> it could be hardcoded in a file and passed to auth-user-pass directive in the config file. Or, even inlined username/password

Well, if the password was somehow "stolen", we could quickly change it on our OpenVPN Server (we are using OPNSense Firewall). With your suggestion, we then needed to do changes on the remote computer (changing the OVPN file with new credentials).

> As 2.7 release is around the corner (hopefully), can you wait for that?

Sure! 😃

**Comment by selvanair** [comment]
> > In OpenVPN 2.7 you can specify `static-challenge "Input OTP" concat`
> 
> [@selvanair](https://github.com/selvanair) do you mean adding this to the OVPN config file and it will show the OTP field on PLAP?

Yes.

That said, concatenated password+OTP is a poor choice. A much better solution is to support original static-challenge protocol in the server-side authentication script/plugin. That will work in 2.6 as well.

**Comment by gitnewcomer** [comment]
> That said, concatenated password+OTP is a poor choice.

I can understand you, but as we (and probably many more) use OpenVPN with OPNSense, this is our only possibility (Password+OTP): https://github.com/opnsense/core/issues/3290

**Comment by schwabe** [comment]
@gitnewcomer that issue litterally is closed 4 years ago and ended with static-challenge being implemented. So you can use static-challenge.

**Comment by gitnewcomer** [comment]
> that issue litterally is closed 4 years ago and ended with static-challenge being implemented. So you can use static-challenge.

@schwabe: this are the settings we use on OPNSense (Local+TOTP), so in the OpenVPN client with PLAP, we are logging in using Password+OTP on the password field. Is there any other way to use Password+OTP on OPNSense by not concatenating Password+OTP on the password field?
From what I understood from @selvanair, the option `concat` from static-challenge will be only available on OpenVPN 2.7... or am I undestanding it wrong?

![Image](https://github.com/user-attachments/assets/f2e592fd-6f68-4892-8751-0f9b8aa74c35)
